File: api/src/new_phone/schemas/hospitality.py
import uuid
from datetime import datetime, time

from pydantic import BaseModel, Field

# --- Room schemas ---


class RoomCreate(BaseModel):
    room_number: str = Field(..., min_length=1, max_length=20)
    extension_id: uuid.UUID | None = None
    floor: str | None = Field(None, max_length=10)
    room_type: str | None = Field(None, max_length=50)
    status: str = "vacant"
    housekeeping_status: str = "clean"
    guest_name: str | None = Field(None, max_length=100)
    guest_checkout_at: datetime | None = None
    wake_up_time: time | None = None
    wake_up_enabled: bool = False
    restricted_dialing: bool = True
    notes: str | None = None


class RoomUpdate(BaseModel):
    room_number: str | None = Field(None, min_length=1, max_length=20)
    extension_id: uuid.UUID | None = None
    floor: str | None = Field(None, max_length=10)
    room_type: str | None = Field(None, max_length=50)
    status: str | None = None
    housekeeping_status: str | None = None
    guest_name: str | None = Field(None, max_length=100)
    guest_checkout_at: datetime | None = None
    wake_up_time: time | None = None
    wake_up_enabled: bool | None = None
    restricted_dialing: bool | None = None
    notes: str | None = None


class RoomResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    tenant_id: uuid.UUID
    room_number: str
    extension_id: uuid.UUID | None
    floor: str | None
    room_type: str | None
    status: str
    housekeeping_status: str
    guest_name: str | None
    guest_checkout_at: datetime | None
    wake_up_time: time | None
    wake_up_enabled: bool
    restricted_dialing: bool
    notes: str | None
    created_at: datetime
    updated_at: datetime


class RoomCheckIn(BaseModel):
    guest_name: str = Field(..., min_length=1, max_length=100)
    guest_checkout_at: datetime | None = None


# --- WakeUpCall schemas ---


class WakeUpCallCreate(BaseModel):
    scheduled_time: datetime


class WakeUpCallResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    tenant_id: uuid.UUID
    room_id: uuid.UUID
    scheduled_time: datetime
    status: str
    attempt_count: int
    created_at: datetime
    updated_at: datetime

File: web/src/api/hospitality.ts
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query"
import { apiClient } from "@/lib/api-client"
import { useAuthStore } from "@/stores/auth-store"
import { queryKeys } from "./query-keys"

export interface Room {
  id: string
  tenant_id: string
  room_number: string
  extension_id: string | null
  floor: string | null
  room_type: string | null
  status: string
  housekeeping_status: string
  guest_name: string | null
  guest_checkout_at: string | null
  wake_up_time: string | null
  wake_up_enabled: boolean
  restricted_dialing: boolean
  notes: string | null
  created_at: string
  updated_at: string
}

export interface RoomCreate {
  room_number: string
  extension_id?: string | null
  floor?: string | null
  room_type?: string | null
  status?: string
  housekeeping_status?: string
  guest_name?: string | null
  guest_checkout_at?: string | null
  wake_up_time?: string | null
  wake_up_enabled?: boolean
  restricted_dialing?: boolean
  notes?: string | null
}

export interface WakeUpCall {
  id: string
  tenant_id: string
  room_id: string
  scheduled_time: string
  status: string
  attempt_count: number
  created_at: string
  updated_at: string
}

export function useRooms(status?: string, floor?: string) {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  return useQuery({
    queryKey: queryKeys.hospitality.rooms(tenantId, status, floor),
    queryFn: () => {
      const params = new URLSearchParams()
      if (status) params.set("room_status", status)
      if (floor) params.set("floor", floor)
      const qs = params.toString()
      return apiClient.get<Room[]>(
        `tenants/${tenantId}/hospitality/rooms${qs ? `?${qs}` : ""}`
      )
    },
    enabled: !!tenantId,
  })
}

export function useRoom(roomId: string) {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  return useQuery({
    queryKey: queryKeys.hospitality.roomDetail(tenantId, roomId),
    queryFn: () =>
      apiClient.get<Room>(`tenants/${tenantId}/hospitality/rooms/${roomId}`),
    enabled: !!tenantId && !!roomId,
  })
}

export function useCreateRoom() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: RoomCreate) =>
      apiClient.post<Room>(`tenants/${tenantId}/hospitality/rooms`, data),
    onSuccess: () =>
      qc.invalidateQueries({
        queryKey: queryKeys.hospitality.all(tenantId),
      }),
  })
}

export function useUpdateRoom() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...data }: { id: string } & Partial<RoomCreate>) =>
      apiClient.patch<Room>(
        `tenants/${tenantId}/hospitality/rooms/${id}`,
        data
      ),
    onSuccess: () =>
      qc.invalidateQueries({
        queryKey: queryKeys.hospitality.all(tenantId),
      }),
  })
}

export function useCheckIn() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      roomId,
      guestName,
      guestCheckoutAt,
    }: {
      roomId: string
      guestName: string
      guestCheckoutAt?: string
    }) =>
      apiClient.post<Room>(
        `tenants/${tenantId}/hospitality/rooms/${roomId}/check-in`,
        {
          guest_name: guestName,
          guest_checkout_at: guestCheckoutAt,
        }
      ),
    onSuccess: () =>
      qc.invalidateQueries({
        queryKey: queryKeys.hospitality.all(tenantId),
      }),
  })
}

export function useCheckOut() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (roomId: string) =>
      apiClient.post<Room>(
        `tenants/${tenantId}/hospitality/rooms/${roomId}/check-out`,
        {}
      ),
    onSuccess: () =>
      qc.invalidateQueries({
        queryKey: queryKeys.hospitality.all(tenantId),
      }),
  })
}

export function useWakeUpCalls(roomId: string) {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  return useQuery({
    queryKey: queryKeys.hospitality.wakeUpCalls(tenantId, roomId),
    queryFn: () =>
      apiClient.get<WakeUpCall[]>(
        `tenants/${tenantId}/hospitality/rooms/${roomId}/wake-up-calls`
      ),
    enabled: !!tenantId && !!roomId,
  })
}

export function useCreateWakeUpCall() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      roomId,
      scheduledTime,
    }: {
      roomId: string
      scheduledTime: string
    }) =>
      apiClient.post<WakeUpCall>(
        `tenants/${tenantId}/hospitality/rooms/${roomId}/wake-up-calls`,
        { scheduled_time: scheduledTime }
      ),
    onSuccess: () =>
      qc.invalidateQueries({
        queryKey: queryKeys.hospitality.all(tenantId),
      }),
  })
}

export function useCancelWakeUpCall() {
  const tenantId = useAuthStore((s) => s.activeTenantId)!
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (wakeUpId: string) =>
      apiClient.post<WakeUpCall>(
        `tenants/${tenantId}/hospitality/wake-up-calls/${wakeUpId}/cancel`,
        {}
      ),
    onSuccess: () =>
      qc.invalidateQueries({
        queryKey: queryKeys.hospitality.all(tenantId),
      }),
  })
}
